src/pkg/systemartifact/manager.go

```diff
@@ -0,0 +1,235 @@
+	cleanupCriteria         map[string]CleanupCriteria
+	lock                    sync.Mutex
+}
+
+func NewManager() Manager {
+	sysArtifactMgr := &systemArtifactManager{
+		regCli:                  registry.Cli,
+		dao:                     dao.NewSystemArtifactDao(),
+		defaultCleanupCriterion: DefaultCleanupCriteria,
+		cleanupCriteria:         make(map[string]CleanupCriteria),
+	}
+	return sysArtifactMgr
+}
+
+func (mgr *systemArtifactManager) Create(ctx context.Context, artifactRecord *model.SystemArtifact, reader io.Reader) (int64, error) {
+
+	id, err := mgr.dao.Create(ctx, artifactRecord)
+
+	if err != nil {
+		return int64(0), err
+	}
+
+	repoName := mgr.getRepositoryName(artifactRecord.Vendor, artifactRecord.Repository)
+	err = mgr.regCli.PushBlob(repoName, artifactRecord.Digest, artifactRecord.Size, reader)
+
+	if err != nil {
+
+		mgr.dao.Delete(ctx, artifactRecord.Vendor, artifactRecord.Repository, artifactRecord.Digest)
+		return int64(0), err
+	}
+	return id, nil
+}
+
+func (mgr *systemArtifactManager) Read(ctx context.Context, vendor string, repository string, digest string) (io.ReadCloser, error) {
+	sa, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return nil, err
+	}
+	repoName := mgr.getRepositoryName(vendor, repository)
+	_, readCloser, err := mgr.regCli.PullBlob(repoName, sa.Digest)
+	if err != nil {
+		return nil, err
+	}
+	return readCloser, nil
+}
+
+func (mgr *systemArtifactManager) Delete(ctx context.Context, vendor string, repository string, digest string) error {
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	err := mgr.regCli.DeleteBlob(repoName, digest)
+
+	if err != nil {
+		return err
+	}
+
+	err = mgr.dao.Delete(ctx, vendor, repository, digest)
+
+	return err
+}
+
+func (mgr *systemArtifactManager) Exists(ctx context.Context, vendor string, repository string, digest string) (bool, error) {
+	_, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return false, err
+	}
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	exist, err := mgr.regCli.BlobExist(repoName, digest)
+
+	if err != nil {
+		return false, err
+	}
+
+	return exist, nil
+}
+
+func (mgr *systemArtifactManager) GetStorageSize(ctx context.Context) (int64, error) {
+	return mgr.dao.Size(ctx)
+}
+
+func (mgr *systemArtifactManager) RegisterCleanupCriteria(vendor string, artifactType string, criteria CleanupCriteria) {
+	key := fmt.Sprintf(keyFormat, vendor, artifactType)
+	defer mgr.lock.Unlock()
+	mgr.lock.Lock()
+	mgr.cleanupCriteria[key] = criteria
+}
+
+func (mgr *systemArtifactManager) GetCleanupCriteria(vendor string, artifactType string) CleanupCriteria {
+	key := fmt.Sprintf(keyFormat, vendor, artifactType)
+	defer mgr.lock.Unlock()
+	mgr.lock.Lock()
+	if criteria, ok := mgr.cleanupCriteria[key]; ok {
+		return criteria
+	}
+	return DefaultCleanupCriteria
+}
+
+func (mgr *systemArtifactManager) Cleanup(ctx context.Context) (int64, int64, error) {
```

Comment:
seems that this method will never return err but nil, do we need this?

Reply:
The current implementation of `Cleanup()` does not return any error. My earlier PR actually had an approach where the cleanup would stop at the first encountered error. However, this would be a problem, since we prevent the deletion of other system artifacts in case any one of the system artifacts could not be deleted (say for e.g. due to an incorrectly implemented cleanup criteria). Hence I only log the error
 `Cleanup()` is an interface method and there could be another implementation provided that would require fail fast and hence the error return.